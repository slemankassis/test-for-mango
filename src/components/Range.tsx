import React, { FC, useEffect, useState, useRef } from 'react';
import InvisibleInput from './InvisibleInput/InvisibleInput';
import RangeControl from './RangeControl';
import { generatePercentages } from '../misc/utils/value-percent-generator';
import { RangeValue } from '../misc/models/RangeValue';
import {
  closestRangeValueByPercentage,
  closestRangeValueByValue,
} from '../misc/utils/closestRangeValue';

enum RangeControls {
  MIN,
  MAX,
}

interface RangeProps {
  options: number[] | { min: number; max: number };
  value?: { min: number; max: number };
  onChange?: (e: { min: number; max: number }) => void;
}

interface RangeState {
  /** Tells if user can edit max and min inputs */
  inputEditable: boolean;
  limitMin: number;
  limitMax: number;
  options: RangeValue[];
  lastActiveControl: RangeControls;
  released: boolean;
  minPos: number;
  maxPos: number;
}

const rangeStateDefault: RangeState = {
  inputEditable: false,
  limitMin: 0,
  limitMax: 0,
  options: [],
  lastActiveControl: RangeControls.MIN,
  released: true,
  minPos: 0,
  maxPos: 100,
};

const Range: FC<RangeProps> = ({ options, value, onChange }) => {
  const rangeBarRef = useRef<HTMLDivElement>(null);

  const [rangeState, setRangeState] = useState<RangeState>({
    ...rangeStateDefault,
  });

  const [val, setVal] = useState<{ min: number; max: number }>({
    min: 0,
    max: 0,
  });

  useEffect(() => {
    let rangeStateTemp: RangeState = { ...rangeStateDefault };

    if (Array.isArray(options)) {
      rangeStateTemp = {
        ...rangeStateTemp,
        options: generatePercentages(options),
      };
    } else if (!isNaN(options.max) && !isNaN(options.min)) {
      // Validate that in fact `min` is minor than `max`
      const min = Math.min(options.min, options.max);
      const max = Math.max(options.min, options.max);
      let steps: number[] = [];

      for (let i = min; i <= max; i++) {
        steps = [...steps, i];
      }

      rangeStateTemp = {
        ...rangeStateTemp,
        options: generatePercentages(steps),
      };
    }

    const limitMin = rangeStateTemp.options[0].value;
    const limitMax = rangeStateTemp.options[rangeStateTemp.options.length - 1].value;
    const min = value?.min ?? limitMin;
    const max = value?.max ?? limitMax;

    setVal({
      min,
      max,
    });

    setRangeState({
      ...rangeStateTemp,
      limitMin,
      limitMax,
      maxPos: closestRangeValueByValue(max, rangeStateTemp.options).percent,
      minPos: closestRangeValueByValue(min, rangeStateTemp.options).percent,
    });
  }, [options, value]);

  const onStartMoving = (el: RangeControls) => {
    setRangeState({
      ...rangeState,
      lastActiveControl: el,
      released: false,
    });
  };

  const onMoving = (
    ev:
      | React.MouseEvent<HTMLDivElement, globalThis.MouseEvent>
      | React.TouchEvent<HTMLDivElement>,
  ) => {
    if (!rangeState.released) {
      if (rangeBarRef.current) {
        const leftSpace = rangeBarRef.current.offsetLeft;
        const rangeWidth = rangeBarRef.current.clientWidth;
        const clientX: number =
          ev.nativeEvent instanceof TouchEvent
            ? (ev as any).touches[0].clientX
            : ev.nativeEvent instanceof MouseEvent
            ? (ev as any).clientX
            : 0;

        const percentajeReal = 100 / (rangeWidth / (clientX - leftSpace));
        let percentaje = percentajeReal;
        if (percentajeReal > 100) {
          percentaje = 100;
        } else if (percentajeReal < 0) {
          percentaje = 0;
        }

        if (rangeState.lastActiveControl === RangeControls.MIN) {
          const minValue = closestRangeValueByPercentage(
            rangeState.minPos,
            rangeState.options,
          );

          setRangeState({
            ...rangeState,
            minPos: percentaje,
          });
          setVal({
            ...val,
            min: minValue.value,
          });
        } else if (rangeState.lastActiveControl === RangeControls.MAX) {
          const maxValue = closestRangeValueByPercentage(
            rangeState.maxPos,
            rangeState.options,
          );

          setRangeState({
            ...rangeState,
            maxPos: percentaje,
          });
          setVal({
            ...val,
            max: maxValue.value,
          });
        }
      }
    }
  };

  const onStopMoving = () => {
    const minValue = closestRangeValueByPercentage(
      rangeState.minPos,
      rangeState.options,
    );
    const maxValue = closestRangeValueByPercentage(
      rangeState.maxPos,
      rangeState.options,
    );

    setRangeState({
      ...rangeState,
      released: true,
    });

    setVal({
      min: minValue.value,
      max: maxValue.value,
    });

    emitValue({
      min: minValue.value,
      max: maxValue.value,
    });
  };

  const emitValue = (val: { min: number; max: number }) => {
    if (onChange) {
      onChange(val);
    }
  };

  return (
    <>
      <InvisibleInput value={val.min} onChange={(ev) => console.log(ev)} />
      <div
        className="range__bar"
        ref={rangeBarRef}
        onMouseMove={(ev) => {
          onMoving(ev);
        }}
        onTouchMove={(ev) => {
          onMoving(ev);
        }}
        onMouseUp={onStopMoving}
        onTouchEnd={onStopMoving}
        onTouchCancel={onStopMoving}
      >
        <RangeControl
          onMouseDown={() => {
            onStartMoving(RangeControls.MIN);
          }}
          onTouchStart={() => {
            onStartMoving(RangeControls.MIN);
          }}
          className="range__bar__control m-min"
          style={{
            zIndex: rangeState.lastActiveControl === RangeControls.MIN ? 1 : 0,
            left: `${rangeState.minPos}%`,
          }}
        />

        <RangeControl
          onMouseDown={() => {
            onStartMoving(RangeControls.MAX);
          }}
          onTouchStart={() => {
            onStartMoving(RangeControls.MAX);
          }}
          className="range__bar__control m-max"
          style={{
            zIndex: rangeState.lastActiveControl === RangeControls.MAX ? 1 : 0,
            left: `${rangeState.maxPos}%`,
          }}
        />
      </div>
      <InvisibleInput value={val.max} onChange={(ev) => console.log(ev)} />
    </>
  );
};

export default Range;
