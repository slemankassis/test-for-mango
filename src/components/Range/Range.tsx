import React, { FC, useEffect, useState, useRef } from 'react';
import InvisibleInput from '../InvisibleInput/InvisibleInput';
import styled from 'styled-components';
import RangeControl from '../RangeControl';
import { generatePercentages } from '../../misc/helpers';
import { RangeValue } from '../../misc/models/RangeValue';
import {
  closestRangeValueByPercentage,
  closestRangeValueByValue,
} from '../../misc/helpers';
import { MinMax } from '../../misc/models/MinMax';

enum RangeControls {
  MIN,
  MAX,
}

function useEvent<K extends keyof WindowEventMap>(
  type: K,
  listener: (this: Window, ev: WindowEventMap[K]) => any,
  options: boolean | AddEventListenerOptions = false,
): void {
  useEffect(() => {
    window.addEventListener(type, listener, options);

    return function cleanup() {
      window.removeEventListener(type, listener);
    };
  });
}

const RangeStyles = styled.div`
  .range {
    display: flex;
    user-select: none;
    padding: 1em 0;

    &.m-disabled {
      position: relative;
      opacity: 0.6;

      &::before {
        content: '';
        position: absolute;
        z-index: 2;
        cursor: not-allowed;
        top: 0;
        right: 0;
        bottom: 0;
        left: 0;
      }

      & > * {
        pointer-events: none;
      }
    }

    &__boundaries {
      max-width: 100%;
    }

    &__bar {
      flex-grow: 1;
      position: relative;
      display: flex;
      flex-direction: column;
      justify-content: center;
      margin: 0 1em;
      min-width: 100px;

      &__control {
        position: absolute;
      }

      &__rail {
        border-radius: 5px;
        height: 5px;
        background-color: #000;
      }
    }
  }
`;

export interface RangeProps {
  options?: number[] | MinMax;
  value?: MinMax;
  onChange?: (e: MinMax) => void;
  unit?: string;
  disabled?: boolean;
}

interface RangeState {
  /** Tells if user can edit max and min inputs */
  inputEditable: boolean;
  limitMin: number;
  limitMax: number;
  options: RangeValue[];
  released: boolean;
  minPos: number;
  maxPos: number;
  disabled: boolean;
  minVal: number;
  maxVal: number;
  lastActiveControl?: RangeControls;
}

const Range: FC<RangeProps> = ({ options, value, onChange, unit, disabled }) => {
  const rangeBarRef = useRef<HTMLDivElement>(null);

  const [rangeState, setRangeState] = useState<RangeState>({
    inputEditable: true,
    limitMin: 0,
    limitMax: 0,
    options: [],
    released: true,
    minPos: 0,
    maxPos: 100,
    disabled: false,
    minVal: 0,
    maxVal: 0,
    lastActiveControl: undefined,
  });

  useEffect(() => {
    let newOptions: RangeValue[];
    let inputEditable = false;

    if (!options) {
      setRangeState((oldRangeState) => {
        return {
          ...oldRangeState,
          disabled: true,
          inputEditable: false,
        };
      });
      return;
    }

    if (Array.isArray(options)) {
      inputEditable = false;
      newOptions = generatePercentages(options);
    } else if (options && !isNaN(options?.max) && !isNaN(options?.min)) {
      // Validate that in fact `min` is minor than `max`
      const min = Math.min(options.min, options.max);
      const max = Math.max(options.min, options.max);
      let steps: number[] = [];

      for (let i = min; i <= max; i++) {
        steps = [...steps, i];
      }

      inputEditable = true;
      newOptions = generatePercentages(steps);
    } else {
      throw new Error('options must be number[] or {min: number, max: number}');
    }

    if (newOptions) {
      const limitMin = newOptions[0]?.value;
      const limitMax = newOptions[newOptions.length - 1]?.value;
      const minVal = value?.min ?? limitMin;
      const maxVal = value?.max ?? limitMax;

      setRangeState((oldRangeState) => {
        return {
          ...oldRangeState,
          limitMin,
          limitMax,
          minVal,
          maxVal,
          maxPos: closestRangeValueByValue(maxVal, newOptions).percent,
          minPos: closestRangeValueByValue(minVal, newOptions).percent,
          inputEditable,
          options: newOptions,
          disabled: !!disabled,
        };
      });
    }
  }, [options, value, disabled]);

  const onStartMoving = (el: RangeControls) => {
    setRangeState((oldRangeState) => {
      return {
        ...oldRangeState,
        released: false,
        lastActiveControl: el,
      };
    });
  };

  const onMoving = (
    ev:
      | React.MouseEvent<HTMLDivElement, globalThis.MouseEvent>
      | React.TouchEvent<HTMLDivElement>,
  ) => {
    if (!rangeState.released) {
      if (rangeBarRef.current) {
        const leftSpace = rangeBarRef.current.offsetLeft;
        const rangeWidth = rangeBarRef.current.clientWidth;
        const clientX: number =
          ev.nativeEvent instanceof TouchEvent
            ? (ev as any).touches[0].clientX
            : ev.nativeEvent instanceof MouseEvent
            ? (ev as any).clientX
            : 0;

        const percentajeReal = 100 / (rangeWidth / (clientX - leftSpace));
        let percentaje = percentajeReal;
        if (percentajeReal > 100) {
          percentaje = 100;
        } else if (percentajeReal < 0) {
          percentaje = 0;
        }

        if (rangeState.lastActiveControl === RangeControls.MIN) {
          if (percentaje > rangeState.maxPos) {
            percentaje = rangeState.maxPos;
          }

          const minValue = closestRangeValueByPercentage(
            percentaje,
            rangeState.options,
          );

          setRangeState((oldRangeState) => {
            return {
              ...oldRangeState,
              minPos: percentaje,
              lastActiveControl: RangeControls.MIN,
              minVal: minValue.value,
            };
          });
        } else if (rangeState.lastActiveControl === RangeControls.MAX) {
          if (percentaje < rangeState.minPos) {
            percentaje = rangeState.minPos;
          }

          const maxValue = closestRangeValueByPercentage(
            percentaje,
            rangeState.options,
          );

          setRangeState((oldRangeState) => {
            return {
              ...oldRangeState,
              maxPos: percentaje,
              lastActiveControl: RangeControls.MAX,
              maxVal: maxValue.value,
            };
          });
        }
      }
    }
  };

  const onStopMoving = () => {
    if (!rangeState.released) {
      const minValue = closestRangeValueByPercentage(
        rangeState.minPos,
        rangeState.options,
      );
      const maxValue = closestRangeValueByPercentage(
        rangeState.maxPos,
        rangeState.options,
      );

      setRangeState((oldRangeState) => {
        return {
          ...oldRangeState,
          released: true,
          minPos: minValue.percent,
          maxPos: maxValue.percent,
          minVal: minValue.value,
          maxVal: maxValue.value,
        };
      });

      emitValue({
        min: minValue.value,
        max: maxValue.value,
      });
    }
  };

  useEvent('mouseup', onStopMoving);

  const emitValue = (val: MinMax) => {
    if (onChange) {
      onChange(val);
    }
  };

  const handleTextInputMin = (ev: React.ChangeEvent<HTMLInputElement>) => {
    let newValue = ev.target.valueAsNumber;

    if (newValue > rangeState.maxVal) {
      newValue = rangeState.maxVal;
    }

    const newRangeValue = closestRangeValueByValue(newValue, rangeState.options);

    setRangeState((oldRangeState) => {
      return {
        ...oldRangeState,
        minPos: newRangeValue.percent,
        minVal: newRangeValue.value,
        lastActiveControl: RangeControls.MIN,
      };
    });

    emitValue({
      max: rangeState.maxVal,
      min: newRangeValue.value,
    });
  };

  const handleTextInputMax = (ev: React.ChangeEvent<HTMLInputElement>) => {
    let newValue = ev.target.valueAsNumber;

    if (newValue < rangeState.minVal) {
      newValue = rangeState.minVal;
    }

    const newRangeValue = closestRangeValueByValue(newValue, rangeState.options);

    setRangeState((oldRangeState) => {
      return {
        ...oldRangeState,
        maxPos: newRangeValue.percent,
        maxVal: newRangeValue.value,
        lastActiveControl: RangeControls.MAX,
      };
    });

    emitValue({
      max: newRangeValue.value,
      min: rangeState.minVal,
    });
  };

  return (
    <RangeStyles>
      <div
        className={`range ${rangeState.disabled ? 'm-disabled' : ''}`}
        onMouseMove={(ev) => {
          onMoving(ev);
        }}
        onTouchMove={(ev) => {
          onMoving(ev);
        }}
        onTouchEnd={onStopMoving}
        onTouchCancel={onStopMoving}
        aria-label="range-selector"
      >
        <InvisibleInput
          value={rangeState.minVal}
          onChange={handleTextInputMin}
          unit={unit}
          disabled={!rangeState.inputEditable || disabled}
          ariaLabel="range input min"
        />
        <div className="range__bar" ref={rangeBarRef}>
          <RangeControl
            aria-label="range control min"
            onMouseDown={() => {
              onStartMoving(RangeControls.MIN);
            }}
            onTouchStart={() => {
              onStartMoving(RangeControls.MIN);
            }}
            className="range__bar__control m-min"
            style={{
              zIndex: rangeState.lastActiveControl === RangeControls.MIN ? 1 : 0,
              left: `${rangeState.minPos}%`,
            }}
          />

          <RangeControl
            aria-label="range control max"
            onMouseDown={() => {
              onStartMoving(RangeControls.MAX);
            }}
            onTouchStart={() => {
              onStartMoving(RangeControls.MAX);
            }}
            className="range__bar__control m-max"
            style={{
              zIndex: rangeState.lastActiveControl === RangeControls.MAX ? 1 : 0,
              left: `${rangeState.maxPos}%`,
            }}
          />

          <div className="range__bar__rail"></div>
        </div>

        <InvisibleInput
          value={rangeState.maxVal}
          onChange={handleTextInputMax}
          unit={unit}
          disabled={!rangeState.inputEditable || disabled}
          ariaLabel="range input max"
        />
      </div>
    </RangeStyles>
  );
};

export default Range;
