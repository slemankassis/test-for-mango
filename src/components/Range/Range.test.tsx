import React from 'react';
import { cleanup, fireEvent, render } from '@testing-library/react';
import TestRenderer from 'react-test-renderer';
import Range, { RangeProps } from './Range';

afterEach(cleanup);

const setup = (props?: RangeProps) => {
  const utils = render(<Range {...props} />);
  const minInput = utils.getByLabelText('range input min');
  const maxInput = utils.getByLabelText('range input max');
  const minControl = utils.getByLabelText('range control min');
  const maxControl = utils.getByLabelText('range control max');
  return {
    minInput,
    maxInput,
    minControl,
    maxControl,
    ...utils,
  };
};

describe('Range snapshots', () => {
  test('snapshot empty range match', () => {
    const tree = TestRenderer.create(<Range />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('snapshot MinMax range match', () => {
    const tree = TestRenderer.create(
      <Range value={{ min: 15, max: 89 }} options={{ min: 1, max: 100 }} />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('snapshot Number[] range match', () => {
    const tree = TestRenderer.create(
      <Range value={{ min: 15, max: 89 }} options={[2, 15, 78, 89]} />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('snapshot disabled range match', () => {
    const tree = TestRenderer.create(<Range disabled={true} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('Range props', () => {
  test('should show the unit', () => {
    const { getAllByText } = setup({ unit: '€' });
    expect(getAllByText(/€/i)[0].textContent).toBe('€');
  });

  test('should choose max and min correctly in number[]', () => {
    const { maxInput, minInput } = setup({
      options: [10, 3, 5, 70, 8],
      value: undefined,
    });

    expect(minInput).toHaveValue(3);
    expect(maxInput).toHaveValue(70);
  });

  test('should choose max and min correctly in reversed MinMax', () => {
    const { maxInput, minInput } = setup({
      options: { min: 100, max: 1 },
      value: undefined,
    });

    expect(minInput).toHaveValue(1);
    expect(maxInput).toHaveValue(100);
  });
});

describe('Range controls', () => {
  test('should min value on top when used', () => {
    const { minControl, maxControl } = setup();

    fireEvent.mouseDown(minControl);

    const styleMinControl = window.getComputedStyle(minControl);
    const styleMaxControl = window.getComputedStyle(maxControl);

    expect(styleMinControl.zIndex).toBe('1');
    expect(styleMaxControl.zIndex).toBe('0');
  });

  test('should max value on top when used', () => {
    const { minControl, maxControl } = setup();

    fireEvent.mouseDown(maxControl);

    const styleMinControl = window.getComputedStyle(minControl);
    const styleMaxControl = window.getComputedStyle(maxControl);

    expect(styleMinControl.zIndex).toBe('0');
    expect(styleMaxControl.zIndex).toBe('1');
  });

  test('should min be at 0% and max at 100% when no value', () => {
    const { minControl, maxControl } = setup();

    const styleMinControl = window.getComputedStyle(minControl);
    const styleMaxControl = window.getComputedStyle(maxControl);

    expect(styleMinControl.left).toBe('0%');
    expect(styleMaxControl.left).toBe('100%');
  });

  test('should update zIndex with alternate clicks', () => {
    const { minControl, maxControl } = setup();

    fireEvent.mouseDown(maxControl);
    fireEvent.mouseDown(minControl);

    const styleMinControl = window.getComputedStyle(minControl);
    const styleMaxControl = window.getComputedStyle(maxControl);

    expect(styleMinControl.zIndex).toBe('1');
    expect(styleMaxControl.zIndex).toBe('0');
  });

  test('should change min value when moves min control', () => {
    const { minControl, minInput } = setup({ options: { min: 20, max: 40 } });

    expect(minInput).toHaveValue(20);

    fireEvent.mouseDown(minControl);
    fireEvent.mouseMove(minControl, { clientX: 100 });
    fireEvent.mouseUp(minControl);

    expect(minInput).toHaveValue(40);

    fireEvent.mouseDown(minControl);
    fireEvent.mouseMove(minControl, { clientX: -1 });
    fireEvent.mouseUp(minControl);

    expect(minInput).toHaveValue(20);
  });

  test('should change max value when moves max control', () => {
    const { maxControl, maxInput } = setup({
      options: { min: 20, max: 40 },
    });

    expect(maxInput).toHaveValue(40);

    fireEvent.mouseDown(maxControl);
    fireEvent.mouseMove(maxControl, { clientX: -100 });
    fireEvent.mouseUp(maxControl);

    expect(maxInput).toHaveValue(20);

    fireEvent.mouseDown(maxControl);
    fireEvent.mouseMove(maxControl, { clientX: 100 });
    fireEvent.mouseUp(maxControl);

    expect(maxInput).toHaveValue(40);
  });

  test('should mouseUp and mouseDown whitout move doesnt change the MIN value', () => {
    const { minControl, minInput } = setup({
      options: { min: 20, max: 40 },
    });

    fireEvent.mouseDown(minControl);
    fireEvent.mouseUp(minControl);

    expect(minInput).toHaveValue(20);
  });

  test('should mouseUp and mouseDown whitout move doesnt change the MAX value', () => {
    const { maxControl, maxInput } = setup({
      options: { min: 20, max: 40 },
    });

    fireEvent.mouseDown(maxControl);
    fireEvent.mouseUp(maxControl);

    expect(maxInput).toHaveValue(40);
  });
});

describe('Range disabled', () => {
  test('should Range be disabled with disabled=true prop', () => {
    const { getByLabelText, minInput, maxInput } = setup({
      options: { min: 1, max: 10 },
      disabled: true,
    });

    expect(getByLabelText('range-selector')).toHaveClass('m-disabled');
    expect(minInput).toBeDisabled();
    expect(maxInput).toBeDisabled();
  });

  test('should has class disabled with no options', () => {
    const { getByLabelText, minInput, maxInput } = setup();

    expect(getByLabelText('range-selector')).toHaveClass('m-disabled');
    expect(minInput).toBeDisabled();
    expect(maxInput).toBeDisabled();
  });
});
