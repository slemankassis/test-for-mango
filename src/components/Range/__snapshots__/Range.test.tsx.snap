// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Range snapshots snapshot MinMax range match 1`] = `
<div
  className="sc-dlfnbm hztomj"
>
  <div
    aria-label="range-selector"
    className="range "
    onMouseMove={[Function]}
    onTouchCancel={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
  >
    <div
      className="sc-bdfBwQ jldiZV"
    >
      <input
        aria-label="range input min"
        className="span-input"
        onBlur={[Function]}
        onChange={[Function]}
        style={
          Object {
            "width": "1ch",
          }
        }
        type="number"
        value="0"
      />
      <span />
    </div>
    <div
      className="range__bar"
    >
      <div
        aria-label="range control min"
        className="sc-gsTCUz iZEpnt range__bar__control m-min"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "left": "14.141414141414142%",
            "zIndex": 0,
          }
        }
      />
      <div
        aria-label="range control max"
        className="sc-gsTCUz iZEpnt range__bar__control m-max"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "left": "88.8888888888889%",
            "zIndex": 0,
          }
        }
      />
      <div
        className="range__bar__rail"
      />
    </div>
    <div
      className="sc-bdfBwQ jldiZV"
    >
      <input
        aria-label="range input max"
        className="span-input"
        onBlur={[Function]}
        onChange={[Function]}
        style={
          Object {
            "width": "1ch",
          }
        }
        type="number"
        value="0"
      />
      <span />
    </div>
  </div>
</div>
`;

exports[`Range snapshots snapshot Number[] range match 1`] = `
<div
  className="sc-dlfnbm hztomj"
>
  <div
    aria-label="range-selector"
    className="range "
    onMouseMove={[Function]}
    onTouchCancel={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
  >
    <div
      className="sc-bdfBwQ jldiZV"
    >
      <input
        aria-label="range input min"
        className="span-input"
        disabled={true}
        onBlur={[Function]}
        onChange={[Function]}
        style={
          Object {
            "width": "1ch",
          }
        }
        type="number"
        value="0"
      />
      <span />
    </div>
    <div
      className="range__bar"
    >
      <div
        aria-label="range control min"
        className="sc-gsTCUz iZEpnt range__bar__control m-min"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "left": "14.942528735632184%",
            "zIndex": 0,
          }
        }
      />
      <div
        aria-label="range control max"
        className="sc-gsTCUz iZEpnt range__bar__control m-max"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "left": "100%",
            "zIndex": 0,
          }
        }
      />
      <div
        className="range__bar__rail"
      />
    </div>
    <div
      className="sc-bdfBwQ jldiZV"
    >
      <input
        aria-label="range input max"
        className="span-input"
        disabled={true}
        onBlur={[Function]}
        onChange={[Function]}
        style={
          Object {
            "width": "1ch",
          }
        }
        type="number"
        value="0"
      />
      <span />
    </div>
  </div>
</div>
`;

exports[`Range snapshots snapshot disabled range match 1`] = `
<div
  className="sc-dlfnbm hztomj"
>
  <div
    aria-label="range-selector"
    className="range m-disabled"
    onMouseMove={[Function]}
    onTouchCancel={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
  >
    <div
      className="sc-bdfBwQ jldiZV"
    >
      <input
        aria-label="range input min"
        className="span-input"
        disabled={true}
        onBlur={[Function]}
        onChange={[Function]}
        style={
          Object {
            "width": "1ch",
          }
        }
        type="number"
        value="0"
      />
      <span />
    </div>
    <div
      className="range__bar"
    >
      <div
        aria-label="range control min"
        className="sc-gsTCUz iZEpnt range__bar__control m-min"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "left": "0%",
            "zIndex": 0,
          }
        }
      />
      <div
        aria-label="range control max"
        className="sc-gsTCUz iZEpnt range__bar__control m-max"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "left": "100%",
            "zIndex": 0,
          }
        }
      />
      <div
        className="range__bar__rail"
      />
    </div>
    <div
      className="sc-bdfBwQ jldiZV"
    >
      <input
        aria-label="range input max"
        className="span-input"
        disabled={true}
        onBlur={[Function]}
        onChange={[Function]}
        style={
          Object {
            "width": "1ch",
          }
        }
        type="number"
        value="0"
      />
      <span />
    </div>
  </div>
</div>
`;

exports[`Range snapshots snapshot empty range match 1`] = `
<div
  className="sc-dlfnbm hztomj"
>
  <div
    aria-label="range-selector"
    className="range "
    onMouseMove={[Function]}
    onTouchCancel={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
  >
    <div
      className="sc-bdfBwQ jldiZV"
    >
      <input
        aria-label="range input min"
        className="span-input"
        onBlur={[Function]}
        onChange={[Function]}
        style={
          Object {
            "width": "1ch",
          }
        }
        type="number"
        value="0"
      />
      <span />
    </div>
    <div
      className="range__bar"
    >
      <div
        aria-label="range control min"
        className="sc-gsTCUz iZEpnt range__bar__control m-min"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "left": "0%",
            "zIndex": 0,
          }
        }
      />
      <div
        aria-label="range control max"
        className="sc-gsTCUz iZEpnt range__bar__control m-max"
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "left": "100%",
            "zIndex": 0,
          }
        }
      />
      <div
        className="range__bar__rail"
      />
    </div>
    <div
      className="sc-bdfBwQ jldiZV"
    >
      <input
        aria-label="range input max"
        className="span-input"
        onBlur={[Function]}
        onChange={[Function]}
        style={
          Object {
            "width": "1ch",
          }
        }
        type="number"
        value="0"
      />
      <span />
    </div>
  </div>
</div>
`;
